task_1_3_1/src/main/java/r/masud/ccfit/nsu/ru/SubStringFinder.java

```diff
@@ -0,0 +1,99 @@
+package r.masud.ccfit.nsu.ru;
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class SubStringFinder {
+
+    /** Note- 00
+     * Prefix-function for Knuth-Morris-Pratt algorithm
+     * @param buffer buffer to calculate prefix-function
+     * @return array of calculated prefix-functions for every position
+     */
+    private static int[] prefixFunction(String buffer) {
+        int[] prefixFuncResult = new int[buffer.length()];
+        prefixFuncResult[0] = 0;
+        for (int i = 1; i < buffer.length(); i++) {
+            int k = prefixFuncResult[i - 1];
+            while ((k > 0) && (buffer.charAt(i) != buffer.charAt(k))) {
+                k = prefixFuncResult[k - 1];
+            }
+            if (buffer.charAt(i) == buffer.charAt(k)) {
+                k++;
+            }
+            prefixFuncResult[i] = k;
+        }
+
+        return prefixFuncResult;
+    }
+
+    /** Note - 01
+     * Function for finding all indexes of first symbol of substring in the buffer.
+     * Using Knuth-Morris-Pratt algorithm
+     * @param buffer buffer to search for substring
+     * @param subString string to find in the buffer
+     * @param shift position of the first symbol of buffer in relation to the file
+     * @return list of indexes of first symbol of found substring in the file
+     */
+    private static List<Integer> findSubStringInBuffer(String buffer, String subString, int shift) {
+        List<Integer> resultList = new ArrayList<>();
+        int[] prefixFuncResult = prefixFunction(subString + "#" + buffer);
+        for (int i = 0; i < buffer.length(); i++) {
+            if (prefixFuncResult[subString.length() + i + 1] == subString.length()) {
+                resultList.add(i - subString.length() + 1 + shift);
+            }
+        }
+
+        return resultList;
+    }
+
+    /** Note -02
+     * Function for finding all substring in file
+     * @param filePath filepath for file to read from
+     * @param subString string to find in file
+     * @return list of indexes of first symbol of found substrings in the file
+     */
+    public static List<Integer> findSubString(String filePath, String subString) {
```

Comment:
Your code should be able to find substring in any type of data: from file, from string in memory, from socket, etc.

See `InputStream`